src/locale.js

```diff
@@ -357,9 +357,14 @@ function parseYear(d, string, i) {
 }
 
 function parseZone(d, string, i) {
-  return /^[+-]\d{4}$/.test(string = string.slice(i, i + 5))
-      ? (d.Z = -string, i + 5) // sign differs from getTimezoneOffset!
-      : -1;
+  var n = /^(Z)|([+-]\d\d)(?:\:?(\d\d))?/.exec(string.slice(i));
```

Comment:
Since the maximum number of possible matching characters here is 6 (`"+12:34"), I think we want `string.slice(i, i + 6)` here for efficiency’s sake.